blogs/WrappedYieldAwaitPattern.md

```diff
@@ -0,0 +1,103 @@
+
+For example, let's consider implementing the LINQ method <a href="http://msdn.microsoft.com/en-us/library/vstudio/bb534804(v=vs.100).aspx">Enumerable.TakeWhile</a>. TakeWhile has two simple conditions to check for: either argument can be null. We can easily implement TakeWhile using yield:
+
+<pre>
+public static IEnumerable<TSource> TakeWhile<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
+{
+    if (source == null) { throw new ArgumentNullException("source"); }
+	if (predicate == null) { throw new ArgumentNullException("predicate"); }
+	
+	foreach (var element in source) 
+	{
+		if (!predicate(element)) { break; }
+		yield return element; 
+	}
+}
+</pre>
+
+However, if we look at <a href="http://referencesource.microsoft.com/#System.Core/System/Linq/Enumerable.cs,28936f7582207b50">Microsoft's implementation of TakeWhile</a>, we see something a bit different:
+
+<pre>
+public static IEnumerable<TSource> TakeWhile<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate) {
+	if (source == null) throw Error.ArgumentNull("source");
+	if (predicate == null) throw Error.ArgumentNull("predicate");
+	return TakeWhileIterator<TSource>(source, predicate);
+}
+
+static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSource> source, Func<TSource, bool> predicate) {
+	foreach (TSource element in source) {
+		if (!predicate(element)) break;
+		yield return element;
+	}
+}
+</pre>
+
+This is what I call the <strong>Wrapped Yield/Await Pattern</strong>: the method is split into a public "wrapper" method that just does argument validation and then calls an inner private method that implements the actual iterator. Why is this done? Let's try running each implementation with invalid arguments:
+
+<pre>
+// MSFT's implementation: this line fails with ArgumentNullException
+var sequence = "abc".TakeWhile(default(Func<char, bool>));
+
+// our implementation: this line succeeds!
+var sequence = "abc".TakeWhile(default(Func<char, bool>));
+
+// ok, now we get the exception
+var array = sequence.ToArray()
+</pre>
+
+What's going on? The key is that the semantics of yield in C# are lazy: none of the code in the iterator block gets executed until the returned sequence is actually enumerated (under the hood, <a href="http://csharpindepth.com/articles/chapter6/iteratorblockimplementation.aspx">yield methods are compiled into state machines</a>). While this is (often) great from a performance perspective, it's less ideal from an error handling perspective, since the exception may occur far from the source of the buggy code (imagine an IEnumerable constructed in one function and then passed around for awhile before being evaluated in another function). The solution is to split the method into two. The outer method does not use yield and is therefore not lazy, so the validation runs right away.
```

Comment:
So the entire function is executed lazily?  Not just the foreach loop ?  I didn't know that and intuitively didn't think that.  I thought everything in the function was executed immediately until the first yield call was hit.  Then execution stopped an was saved for later.